app-libs/stf/src/stf_sgx.rs

```diff
@@ -152,23 +166,171 @@ where
 
 	fn execute_call(
 		state: &mut State,
+		shard: &ShardIdentifier,
 		call: TCS,
 		calls: &mut Vec<ParentchainCall>,
 		node_metadata_repo: Arc<NodeMetadataRepository>,
 	) -> Result<(), Self::Error> {
-		state.execute_with(|| call.execute(calls, node_metadata_repo))
+		state.execute_with(|| call.execute(calls, shard, node_metadata_repo))
 	}
 
-	fn on_initialize(state: &mut State, now: Moment) -> Result<(), Self::Error> {
+	fn on_initialize(
+		state: &mut State,
+		shard: &ShardIdentifier,
+		integritee_block_number: BlockNumber,
+		now: Moment,
+	) -> Result<(), Self::Error> {
 		trace!("on_initialize called at epoch {}", now);
 		state.execute_with(|| {
 			// as pallet_timestamp doesn't export set_timestamp in no_std, we need to re-build the same behaviour
 			sp_io::storage::set(&storage_value_key("Timestamp", "Now"), &now.encode());
 			sp_io::storage::set(&storage_value_key("Timestamp", "DidUpdate"), &true.encode());
 			<Runtime::OnTimestampSet as OnTimestampSet<_>>::on_timestamp_set(now.into());
+			// ensure we're assuming the correct storage encoding based on pallet version
+			if get_mirrored_parentchain_storage_by_key_hash::<u16>(
+				<EnclaveBridgeStorage as EnclaveBridgeStorageKeys>::pallet_version(),
+				&ParentchainId::Integritee,
+			)
+			.map_or(false, |v| v <= 1)
+			{
+				if let Some(config) = get_mirrored_parentchain_storage_by_key_hash(
+					EnclaveBridgeStorage::upgradable_shard_config(shard),
+					&ParentchainId::Integritee,
+				) {
+					ita_sgx_runtime::ShardManagementCall::<ita_sgx_runtime::Runtime>::set_shard_config {
+						config,
+						parentchain_block_number: integritee_block_number,
+					}
+						.dispatch_bypass_filter(ita_sgx_runtime::RuntimeOrigin::root())
+						// Replace with `inspect_err` once it's stable.
+						.map_err(|_| {
+							error!("Failed to mirror shard config");
+						})
+						.ok();
+				} else {
+					warn!("mirrored state not available: shard_config");
+				}
+			} else {
+				warn!("Parentchain Integritee pallet version mismatch. Can't sync mirrored state safely");
+			}
 		});
 		Ok(())
 	}
+
+	fn maintenance_mode_tasks(
+		state: &mut State,
+		shard: &ShardIdentifier,
+		integritee_block_number: BlockNumber,
+		calls: &mut Vec<ParentchainCall>,
+		node_metadata_repo: Arc<NodeMetadataRepository>,
+	) -> Result<(), Self::Error> {
+		state.execute_with(|| {
+			let maintenance_mode_age = ShardManagement::upgradable_shard_config()
+				.map(|(_config, updated_at)| updated_at)
+				.unwrap_or(integritee_block_number);
+			if maintenance_mode_age
+				>= MinimalChainSpec::maintenance_mode_duration_before_retirement(
+					shielding_target_genesis_hash().unwrap_or_default(),
```

Comment:
Does default make sense here??

Reply:
the fn will need a fallback behavior. We could let the arg be an Option instead, but it doesn't feel right to make the arg optional